### PR TITLE
Support reset password for vm

### LINKED ISSUE #979: Support reset password for vm
/kind feature
 in the current version,after launch a cluter,if we do not know the password for root from the image ,we can only connet to the vm by keypair,I hope we can reset the password,the we could connect to the vm by password

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by ruifaling**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #979

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

File: api/v1alpha4/openstackmachine_types.go [hunk 1]
```diff
@@ -90,6 +90,8 @@ type OpenStackMachineSpec struct {
 	// IdentityRef is a reference to a identity to be used when reconciling this cluster
 	// +optional
 	IdentityRef *OpenStackIdentityReference `json:"identityRef,omitempty"`
+
+	AdminPass string `json:"adminPass,omitempty"`
```
Comment (apricote) [inline api/v1alpha4/openstackmachine_types.go:94]: Having this password in the Resources might conflict on some security requirements. What do you think about reading the password from a secret, similar to `OpenStackCluster.Spec.IdentityRef`?

Comment (ruifaling) [reply api/v1alpha4/openstackmachine_types.go:94]: Thanks,this might be a good suggestion

File: controllers/openstackmachine_controller.go [hunk 2]
```diff
@@ -469,7 +470,16 @@ func (r *OpenStackMachineReconciler) getBootstrapData(ctx context.Context, machi
...
+	//return base64.StdEncoding.EncodeToString(value), nil
+	var buffer bytes.Buffer
+	buffer.Write(value)
+	if openStackMachine.Spec.AdminPass != ""{
+		adminPass := openStackMachine.Spec.AdminPass
+		resetpwd := "  - '(echo "+ adminPass +";sleep 1; echo "+ adminPass +") | passwd root' "
+		userPwd := []byte(resetpwd)
+		buffer.Write(userPwd)
+	}
+	return base64.StdEncoding.EncodeToString(buffer.Bytes()), nil
```
Comment (mdbooth) [inline controllers/openstackmachine_controller.go:482]: I don't think this is the right place for this. I think instead you need to look into how to have your AdminPass added directly to the bootstrap data.

Comment (jichenjc) [reply controllers/openstackmachine_controller.go:482]: exactly, I think this kind of info should be doable through openstack userdata etc?
add admin pwd here seems not a good place here

not sure https://docs.openstack.org/api-ref/compute/?expanded=change-administrative-password-changepassword-action-detail#change-administrative-password-changepassword-action
this is helpful ?

**Comment by ruifaling** [comment]
>

**Comment by ruifaling** [comment]
@jichenjc ,hi jichen，could you review the codes and help me to find out why it is always faild in the step of make generate ?

**Comment by tobiasgiese** [comment]
@ruifaling you have to run `make generate` locally and commit these changes. The test is just validating if there is a diff after generating the code

**Comment by jichenjc** [comment]
yeah, as @tobiasgiese suggested, check
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1081/pull-cluster-api-provider-openstack-test/1470648459680813056/build-log.txt

and run `make generate` to see whether you have additional changes in your latest commit
if so, please commit them then submit again

**Comment by ruifaling** [comment]
> @ruifaling you have to run `make generate` locally and commit these changes. The test is just validating if there is a diff after generating the code

Thank you,thank you very much

**Comment by ruifaling** [comment]
> yeah, as @tobiasgiese suggested, check https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1081/pull-cluster-api-provider-openstack-test/1470648459680813056/build-log.txt
> 
> and run `make generate` to see whether you have additional changes in your latest commit if so, please commit them then submit again

Thanks

**Comment by mdbooth** [comment]
@ruifaling But please bear in mind that I don't think we are likely to accept this change even if the tests pass: CAPO really isn't concerned with the contents of UserData and I don't think we want to add support for modifying it in flight.

I wish I could point you to the component which generates UserData, but unfortunately I haven't had to modify it myself yet so I don't know. I do know how OpenShift generates it, but that's quite different to how CAPI does it. I suspect others here do know, though! Can anybody help out?

**Comment by ruifaling** [comment]
> @ruifaling But please bear in mind that I don't think we are likely to accept this change even if the tests pass: CAPO really isn't concerned with the contents of UserData and I don't think we want to add support for modifying it in flight.
> 
> I wish I could point you to the component which generates UserData, but unfortunately I haven't had to modify it myself yet so I don't know. I do know how OpenShift generates it, but that's quite different to how CAPI does it. I suspect others here do know, though! Can anybody help out?

In the openstack compute api: [https://docs.openstack.org/api-ref/compute/?expanded=change-administrative-password-changepassword-action-detail,create-server-detail](url)  here is some example of create server with adminPass,and if you do not want accept it ,after the tests pass ,I will close the pr.Thanks  for your advice.

**Comment by ruifaling** [comment]
> @ruifaling But please bear in mind that I don't think we are likely to accept this change even if the tests pass: CAPO really isn't concerned with the contents of UserData and I don't think we want to add support for modifying it in flight.
> 
> I wish I could point you to the component which generates UserData, but unfortunately I haven't had to modify it myself yet so I don't know. I do know how OpenShift generates it, but that's quite different to how CAPI does it. I suspect others here do know, though! Can anybody help out?



> > @ruifaling But please bear in mind that I don't think we are likely to accept this change even if the tests pass: CAPO really isn't concerned with the contents of UserData and I don't think we want to add support for modifying it in flight.
> > I wish I could point you to the component which generates UserData, but unfortunately I haven't had to modify it myself yet so I don't know. I do know how OpenShift generates it, but that's quite different to how CAPI does it. I suspect others here do know, though! Can anybody help out?
> 
> In the openstack compute api: [https://docs.openstack.org/api-ref/compute/?expanded=change-administrative-password-changepassword-action-detail,create-server-detail](url) here is some example of create server with adminPass,and if you do not want accept it ,after the tests pass ,I will close the pr.Thanks for your advice.

admin
@mdbooth

**Comment by jichenjc** [comment]
@ruifaling yes, we knew there is API that is able to do this ,but it's not the desired way for your use case

I think you might want to check 
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/tasks/kubeadm-bootstrap.md#how-does-cabpk-work

as it says, you can update the KubeadmConfig to cloud-init script which you can inject your password update throught cloud-init spec, e.g 

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1081/pull-cluster-api-provider-openstack-e2e-test/1470931830566817792/artifacts/clusters/bootstrap/resources/e2e-vp4l65/KubeadmConfig/cluster-e2e-vp4l65-control-plane-5dnwk.yaml
 
is what our test is running, I think you can follow the guide above to see how we can achieve this ..

**Comment by ruifaling** [comment]
@jichenjc https://github.com/kubernetes-sigs/cluster-api/pull/5885; is this？

**Comment by jichenjc** [comment]
@ruifaling that might be the case, need read more and comment on that PR directly. thanks~

**Pull request closed without merging**